README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
ramlegacy
=========

[![Travis Build Status](https://travis-ci.com/ropensci/ramlegacy.svg?branch=master)](https://travis-ci.com/ropensci/ramlegacy) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/ropensci/ramlegacy?branch=master&svg=true)](https://ci.appveyor.com/project/kshtzgupta1/ramlegacy) [![Coverage status](https://codecov.io/gh/ropensci/ramlegacy/branch/master/graph/badge.svg)](https://codecov.io/github/ropensci/ramlegacy) [![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip) [![](https://badges.ropensci.org/264_status.svg)](https://github.com/ropensci/software-review/issues/264)

-   **Authors**: Kshitiz Gupta, [Carl Boettiger](http://www.carlboettiger.info/)
-   **License**: [MIT](http://opensource.org/licenses/MIT)
-   [Package source code on Github](https://github.com/ropensci/ramlegacy)
-   [**Submit Bugs and feature requests**](https://github.com/ropensci/ramlegacy/issues)

`ramlegacy` is an R package that supports caching and reading in different versions of the RAM Legacy Stock Assessment Data Base, an online compilation of stock assessment results for commercially exploited marine populations from around the world. More information about the database can be found [here.](https://www.ramlegacy.org/)

What does `ramlegacy` do?
-------------------------

-   Provides a function `download_ramlegacy()`, to download all the available versions of the RAM Legacy Stock Assessment Excel Database and cache them on the user's computer as serialized RDS objects. This way once a version has been downloaded it doesn't need to be re-downloaded for subsequent analysis.
-   Supports reading in specified tables or all tables from a cached version of the database through a function `load_ramlegacy()`
-   Provides a function `ram_dir()` to view the path of the location where the downloaded database was cached.

Installation
------------

You can install the development version from [github](https://github.com/ropensci/ramlegacy) with:

``` r
install.packages("devtools")
library(devtools)
install_github("ropensci/ramlegacy")
```

To ensure that the vignette is installed along with the package make sure to remove `--no-build-vignettes` from the `build_opts` in `install_github`

Usage
-----

Please see the ramlegacy [vignette](https://ropensci.github.io/ramlegacy/articles/ramlegacy.html) for more detailed examples and additional package functionality.

Start by loading the package using `library`.

``` r
library(ramlegacy)
```

### download\_ramlegacy

`download_ramlegacy()` downloads the specified version of **RAM Legacy Stock Assessment Excel Database** and then saves it as an RDS object in user’s application data directory as detected by the [rappdirs](https://CRAN.R-project.org/package=rappdirs) package. This location is also where `load_ramlegacy()` by default will look for the downloaded database.

``` r
# downloads version 4.44
download_ramlegacy(version = "4.44")
```

If version is not specified then `download_ramlegacy` defaults to downloading current latest version (4.44) :

``` r
# downloads current latest version 4.44
download_ramlegacy()
```

The latest versions of the RAM Legacy Database are [archived in Zenodo](https://zenodo.org/communities/rlsadb/) but the older versions (v4.3, v3.0, v2.5, v2.0, v1.0) are not. To ensure access to these older versions of the database `download_ramlegacy` supports downloading them from this [github repository](https://www.github.com/kshtzgupta1/ramlegacy-assets/):

``` r
# downloads older version 4.3
download_ramlegacy(version = "4.3")
```

### load\_ramlegacy

After the specified version of the database has been downloaded and cached on your local machine through `download_ramlegacy` you can call `load_ramlegacy` to obtain a list of specific tables/all the tables from that version of the database. If version is not specified but tables is then `load_ramlegacy` defaults to returning a list containing the specified dataframes from the latest version (currently 4.44). If both version and tables are not specified then `load_ramlegacy` defaults to returning a list containing all the dataframes in the latest version (currently 4.44)

``` r
# get a list containing area and bioparams tables from
# version 4.3 of the database
load_ramlegacy(version = "4.3", tables = c("area", "bioparams"))

# get a list containing area and bioparams tables from version 4.44
# of the database
load_ramlegacy(version = "4.44", tables = c("area", "bioparams"))

# if tables is specified but version is not then the function defaults
# to returning a list containing the specified tables from the current
# latest version 4.44
load_ramlegacy(tables = c("area", "bioparams"))

# since both tables and version are not specified the function returns
# a list containing all the tables from the current latest version 4.44
load_ramlegacy()
```

To learn more about the different tables present in the database, what the various acronyms mean and the different stock summaries accompanying the databases please see this [page.](https://ropensci.github.io/ramlegacy/articles/tables_description.html)

### ram\_dir

To view the exact path where a certain version of the database was downloaded and cached by `download_ramlegacy` you can run `ram_dir(vers = 'version')`, specifying the version number inside the function call:

``` r
# download version 4.44
download_ramlegacy(version = "4.44")

# view the location where version 4.44 of the database was
# downloaded and cached
ram_dir(vers = "4.44")
```

Similar Projects
----------------

1.  [`ramlegacy`](https://github.com/seananderson/ramlegacy) Sean Anderson has a namesake package that appears to be a stalled project on GitHub (last updated 9 months ago). However, unlike this package which supports downloading and reading in the Excel version of the database, Sean Anderson's project downloads the Microsoft Access version and converts it to a local sqlite3 database.

2.  [`RAMlegacyr`](https://github.com/ashander/RAMlegacyr) `RAMlegacyr` is an older package last updated in 2015. Similar to Sean Anderson's project, the package seems to be an R interface for the Microsoft Access version of the RAM Legacy Stock Assessment Database and provides a set of functions using RPostgreSQL to connect to the database.

Citation
--------

Current and older versions of the RAM Legacy Database are [archived in Zenodo](https://zenodo.org/communities/rlsadb/), each version with its own unique DOI. The suggested format for citing data is:

RAM Legacy Stock Assessment Database. 2018. Version 4.44-assessment-only. Released 2018-12-22. Accessed \[Date accessed YYYY-MM-DD\]. Retrieved from [DOI:10.5281/zenodo.2542919.](https://zenodo.org/record/2542919#.XE-rFs9KjBI)

The primary publication describing the RAM Legacy Stock Assessment Database, and suggested citation for general use is:

Ricard, D., Minto, C., Jensen, O.P. and Baum, J.K. (2012) Evaluating the knowledge base and status of commercially exploited marine species with the RAM Legacy Stock Assessment Database. Fish and Fisheries 13 (4) 380-398. [DOI: 10.1111/j.1467-2979.2011.00435.x](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1467-2979.2011.00435.x)

Several [publications](http://sites.uw.edu/ramlegac/publications/) have relied on the RAM Legacy Stock Assessment Database.

[![ropensci\_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
NEWS 
====

For more fine-grained list of changes or to report a bug, consult 

* [The issues log](https://github.com/ropensci/ramlegacy/issues)
* [The commit log](https://github.com/ropensci/ramlegacy/commits/master)

Versioning
----------

Releases will be numbered with the following semantic versioning format:

`<major>.<minor>.<patch>`

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor 
  and patch)
* New additions without breaking backward compatibility bumps the minor 
  (and resets the patch)
* Bug fixes and misc changes bumps the patch
* Following the RStudio convention, a .99 is appended after the patch
  number to indicate the development version on Github.  Any version
  Coming from Github will now use the .99 extension, which will never
  appear in a version number for the package on CRAN. 


v0.1.0
------
* Initial Release

v0.2.0
-----
* Removed the loading behavior of `library(ramlegacy)` leaving the loading solely to `load_ramlegacy`
* Instead of assigning all the tables to global environment `load_ramlegacy` now just returns a list of tables that can be specified through `tables` argument.
* Modified `download_ramlegacy` and `load_ramlegacy` so that that while the default is still to download and read in the dataframes from the rappdirs directory the functions now also support downloading to a location chosen by the user and reading from that location. 



R/load_ramlegacy.R
#' @name load_ramlegacy
#' @family ramlegacy functions
#' @title Read-in dataframes from downloaded RAM Legacy Database
#' @description Returns a list of specific dataframes or a list of all the
#'  dataframes present in the requested version of the database.
#' @param version A character vector of length 1 specifying the version number
#' of the database. As of writing this package, the available versions are
#' "1.0", "2.0", "2.5", "3.0", "4.3", "4.40", "4.41" and "4.44". If version
#' argument is not specified then it defaults to newest version (v4.44).
#' @param tables A character vector specifying the names of particular
#' dataframes to load from a specified version. If not specified then
#' a list containing all the dataframes within the requested version
#' is returned. For a description of the different tables present
#' in the database please see below.
#' @param ram_path path to the local directory where the specified version of
#' the RAM Legacy Stock Excel Assessment Database was downloaded. By default
#' this path is set to within the rappdirs directory and can be viewed using
#' calling the function \code{\link{ram_dir}} and specifying the version number
#' inside the function call. Although this is not the \strong{recommended}
#' approach \code{load_ramlegacy} supports reading in the database's
#' dataframes from a user-specified path provided that the database is present
#' at the specified path as an rds object.
#' @export
#'
#' @examples
#' \donttest{
#' \dontshow{Sys.setenv(RAM_HOME = tempfile())}
#' # first download version 4.44 of the database
#' download_ramlegacy(version = "4.44")
#'
#' # get a list containing area and bioparams tables
#' # from version 4.44 database
#' load_ramlegacy(version = "4.44", tables = c("area", "bioparams"))
#' }
#' @section Description of the dataframes present in the database:
#'
#' \itemize{
#' \item metadata: Table with summarized metadata (only available in newer
#'  versions 4.40, 4.41, 4.44)
#' \item stock: This stores the stock database table
#' \item assessment: This stores the assessment database table
#' \item taxonomy: This stores the taxonomy database table
#' \item management: This stores the management database table
#' \item assessor: This stores the assessor database table
#' \item assessmetod: This stores the assessmetod database table
#' \item area: This stores the area database table
#' \item biometrics: This stores the biometrics database table
#' \item tsmetrics: This stores the tsmetrics database table
#' \item timeseries: The time series data is a matrix object with the following
#' headers/columns: (1) assessid (2) stockid (3) stocklong (4) tsid (5) tsyear
#' (6) tsvalue
#' \item bioparams: The parameter data is a matrix object with the following
#' headers/columns: (1) assessid (2) stockid (3) stocklong (4) bioid (5) biovalue
#' (6) bioyear (7) bionotes
#' \item timeseries_values_views: This stores the timeseries values with timeseries
#' type along the columns and stocks along the rows
#' \item timeseries_units_views: This stores the timeseries values with timeseries
#' type along the columns and stocks along the rows
#' \item timeseries_ids_views: This stores the timeseries IDs with timeseries type
#'  along the columns and stocks along the rows
#' \item timeseries_assessments_views: This stores the timeseries assessments with
#' timeseries type along the columns and stocks along the rows
#' \item timeseries_notes_views: This stores the timeseries notes with timeseries
#' type along the columns and stocks along the rows
#' \item timeseries_sources_views: This stores the timeseries sources with timeseries
#' type along the columns and stocks along the rows
#' \item timeseries_years_views: This stores the timeseries years with timeseries
#' type along the columns and stocks along the rows
#' \item bioparams_values_views: This stores the reference point values, with
#' reference point type along the columns and stocks along the rows
#' \item bioparams_units_views: This stores the reference point units, with
#' reference point type along the columns and stocks along the rows
#' \item bioparams_ids_views: This stores the reference point IDs, with reference
#' point type along the columns and stocks along the rows
#' \item bioparams_assessments_views: This stores the reference point assessments,
#' with reference point type along the columns and stocks along the rows
#' \item bioparams_sources_views: This stores the reference point sources, with
#' reference point type along the columns and stocks along the rows
#' \item bioparams_notes_views: This stores the reference point notes, with
#' reference point type along the columns and stocks along the rows
#' }
#'
#' @section v4.40, v4.41, and v4.44 also contain tables for individual most-used time series:
#' \itemize{
#' \item tb.data: Total Biomass
#' \item ssb.data: Spawning Stock Biomass
#' \item tn.data: Total Abundance
#' \item r.data: Recruits
#' \item tc.data: Total Catch
#' \item tl.data: Total Landings
#' \item recc.data: Recreational Catch
#' \item f.data: Fishing Mortality
#' \item er.data: Exploitation Rate
#' \item divtb.data: TB/TBmsy
#' \item divssb.data: SSB/SSBmsy
#' \item ivf.data: F/Fmsy
#' \item diver.data: ER/ERmsy
#' \item divbpref.data: B/Bmsypref
#' \item divupref.data: U/Umsypref
#' \item tbbest.data: TBbest
#' \item tcbest.data: TCbest
#' \item erbest.data: ERbest
#' \item divtb.mgt.data: TB/TBmgt
#' \item divssb.mgt.data: SSB/SSBmgt
#' \item divf.mgt.data: F/Fmgt
#' \item diver.mgt.data: ER/ERmgt
#' \item divbpref.mgt.data: B/Bmgtpref
#' \item divupref.mgt.data: U/Umgtpref
#' \item cpair.data: Cpair
#' \item tac.data: TAC
#' \item cadv.data: Cadvised
#' \item survb.data: survB
#' \item cpue.data: CPUE
#' \item effort.data: EFFORT
#' }

load_ramlegacy <- function(version = NULL, tables = NULL, ram_path = NULL) {
  ram_url <- "https://doi.org/10.5281/zenodo.2542918"

  # convert names to lowercase

  if (!is.null(version)) {
    # ensure that the version is properly formatted
    version <- fmt_version(version)
    check_version_arg(version)
  } else {
    version <- find_latest(ram_url)
  }

  if (is.null(ram_path)) {
    vers_path <- file.path(ram_dir(), version)
    if (version < "4.40") {
      rds_path <- file.path(vers_path, paste0("v", version, ".rds"))
    } else {
      rds_path <- file.path(vers_path, paste0("RLSADB v", version))
      rds_path <- file.path(rds_path, "DB Files With Assessment Data")
      rds_path <- file.path(rds_path, paste0("v", version, ".rds"))
    }
  } else {
    check_path(ram_path)
    rds_path <- ram_path
  }
  # make sure version is present
  if (!file.exists(rds_path)) {
    stop(paste0("Version ", version, " not found locally."))
  }

  list_dataframes <- readRDS(rds_path)

  if (!is.null(tables)) {
    tables <- tolower(tables)
    listToReturn <- vector("list", length(tables))
    for (i in seq_len(length(tables))) {
      #  construct df name
      df_name <- tables[i]
      if (grepl("\\.data", tables[i])) {
        most.used.time.series <- list_dataframes[[26]]
        listToReturn[[i]] <- most.used.time.series[[df_name]]
      } else {
        listToReturn[[i]] <- list_dataframes[[df_name]]
      }
    }

    names(listToReturn) <- tables
    class(listToReturn) <- c("ramlist", class(listToReturn))
    return(listToReturn)
  } else {
    class(list_dataframes) <- c("ramlist", class(list_dataframes))
    return(list_dataframes)
  }
}
